tutorial: Using AWS ALB with Azure AD to authenticate requests (#5216)
backstage/backstage: contrib/docs/tutorials/aws-alb-aad-oidc-auth.md

> # Using AWS Application Load Balancer with Azure Active Directory to authenticate requests

Reviewer: For this to show up properly on the microsite you'll need to add some frontmatter and fill in `microsite/sidebars.json`. You can preview it using `yarn start` in `./microsite`

Author: Thanks @Rugvip 
Other tutorials under `contrib/docs/tutorial` don't have this, would you still recommend adding it?

Reviewer 2: This should only be added for things in `/docs` afaik.

Author: @Rugvip / @timbonicus is there anything else that needs to be done before this PR could be merged? Thank you 🙏

> Backstage allows to offload the responsibility of authenticating users to AWS Application Load Balancer (**ALB**), leveraging the authentication support on ALB.

Reviewer 2: ```suggestion
Backstage allows offloading the responsibility of authenticating users to an AWS Application Load Balancer (**ALB**), leveraging the authentication support on ALB.
```

This tutorial shows how to use authentication on an ALB sitting in front of Backstage.
Azure Active Directory (**AAD**) is used as identity provider but any identity provider supporting OpenID Connect (OIDC) can be used.

It is assumed an ALB is already serving traffic in front of a Backstage instance configured to serve the frontend app from the backend.

## Infrastructure setup

### AAD App

The AAD App is used to execute the authentication flow, serve and refresh the identity token.

Create the AAD App following the steps outlined in `Create a Microsoft App Registration in Microsoft Portal` section from the tutorial [Monorepo App Setup With Authentication][monorepo-app-setup-with-auth].

Instead of `localhost` addresses, use the following values.

- Identifier URI: `https://backstage.yourdomain.com`
- Redirect URI: `https://backstage.yourdomain.com/oauth2/idpresponse`

`Application (client) Id`, `Directory (tenant) ID` and `client secret`values will be used while configuring the ALB.

### ALB

> In AWS console configure ALB Authentication as described below:

Reviewer 2: ```suggestion
In the AWS console, configure ALB Authentication:
```

- Edit the ALB rule used to forward the traffic to Backstage and add a new `Authenticate` action. The action will have higher priority compared to the existing `Forward to`.
- Select `OIDC` under `Authenticate`
- Set `Issuer` to `https://login.microsoftonline.com/{TENANT_ID}/v2.0`, replacing `{TENANT_ID}` with the value of `Directory (tenant) ID` of the AAD App.
- Set `Authorization endpoint` to `https://login.microsoftonline.com/{TENANT_ID}/oauth2/v2.0/authorize`, replacing `{TENANT_ID}` with the value of `Directory (tenant) ID` of the AAD App.
- Set `Token endpoint` to `https://login.microsoftonline.com/{TENANT_ID}/oauth2/v2.0/authorize`, replacing `{TENANT_ID}` with the value of `Directory (tenant) ID` of the AAD App.
- Set `User info endpoint` to `https://graph.microsoft.com/oidc/userinfo`
- Set `Client ID` to the AAD App `Application (client) Id`
- Set `Client secret` to the AAD APP `client secret`

Use the following advanced settings:

- `Session cookie name` = `AWSELBAuthSessionCookie`
- `Session timeout` = `604800` seconds
- `Scope` = `openid profile offline_access`
- `Action on unauthenticated request` = `Autenticate (client reattempt)`

> Once you've saved the action, you should see an authentication flow be triggered against AAD when visiting Backstage address at `https://backstage.yourdomain.com`. The flow will not complete successfully as Backstage app isn't yet configured properly.

Reviewer 2: ```suggestion
Once you've saved the action, you should see an authentication flow be triggered against AAD when visiting Backstage address at `https://backstage.yourdomain.com`. The flow will not complete successfully as the Backstage app isn't yet configured properly.
```

## Backstage changes

### Frontend

> The Backstage App needs a SingInPage when authentication is required.

Reviewer 2: ```suggestion
The Backstage App needs a SignInPage when authentication is required.
```

When using ALB authentication Backstage will only be loaded once the user has successfully authenticated; we won't need to display a SignIn page, however we will need to create a dummy SignIn component that can refresh the token.

- edit `packages/app/src/App.tsx`

> - import the following two additional definitions from `@backstage/core`: `useApi`, `configApiRef`; these will be used to check wether backstage is running locally or behind an ALB

Reviewer 2: ```suggestion
- import the following two additional definitions from `@backstage/core`: `useApi`, `configApiRef`; these will be used to check whether Backstage is running locally or behind an ALB
```

- add the following definition just before the app is created (`const app = createApp`):

```ts
const DummySignInComponent: any = (props: any) => {
  const config = useApi(configApiRef);
  const shouldAuth = !!config.getOptionalConfig('auth.providers.awsalb');
  if (shouldAuth) {
    fetch(`${window.location.origin}/api/auth/awsalb/refresh`)
      .then(data => data.json())
      .then(data => {
        props.onResult({
          userId: data.backstageIdentity.id,
          profile: data.profile,
        });

>       });

Reviewer: Would be best to catch and render any errors as well. Since this is just an example it should be fine to just set some error state and throw it in the div.

  } else {
    // when running locally we default user identity to `Local User`
    props.onResult({

>       userId: 'local user',

Reviewer: What do you think about sticking to the `'guest'` user convention here? https://github.com/backstage/backstage/blob/ae7b2a28d60159b94eda7020b41e7c3a86c8fb3a/packages/core-api/src/app/App.tsx#L400

      profile: {
        email: 'local.user@yourdomain.com',
        displayName: 'Local User',
        picture: '',
      },
    });
  }

  return <div />;
};
```

> - use `DummySingInComponent` as `SignInPage` by changing `createApp` statement as follow:

Reviewer 2: ```suggestion
- use `DummySignInComponent` as `SignInPage` by changing the `createApp` call:
```

```ts
const app = createApp({
  apis,
  plugins: Object.values(plugins),
  components: {
    SignInPage: DummySignInComponent,
  },

>   bindRoutes({ bind }) {

Reviewer: I'd say omit this and just highlight the addition to `components` instead. Something like 

```ts
const app = createApp({
  ...
  components: {
    SignInPage: DummySignInComponent,
    ...
  },
  ...
})
```

    bind(catalogPlugin.externalRoutes, {
      createComponent: scaffolderPlugin.routes.root,
    });
    bind(apiDocsPlugin.externalRoutes, {
      createComponent: scaffolderPlugin.routes.root,
    });
  },
});
```

### Backend

> When using ALB auth it is not possible to leverage the built-in auth config discovery mechanism implemented in the app created by default: a bespoke logic needs to be implemented.

Reviewer 2: ```suggestion
When using ALB auth it is not possible to leverage the built-in auth config discovery mechanism implemented in the app created by default; bespoke logic needs to be implemented.
```
